torch/csrc/jit/tensorexpr/ir_simplifier.cpp

```diff
@@ -2910,6 +2911,152 @@ ExprPtr SimplifierUnderContext::mutate(DivPtr v) {
   return alloc<Div>(lhs_new, rhs_new);
 }
 
+ExprPtr SimplifierUnderContext::mutate(IfThenElsePtr v) {
+  ExprPtr condition = v->condition();
+  ExprPtr true_val = v->true_value();
+  ExprPtr false_val = v->false_value();
+
+  auto simplified_condition =
+      IRSimplifier::simplify(condition->accept_mutator(this));
+  auto simplified_true_val =
+      IRSimplifier::simplify(true_val->accept_mutator(this));
+  auto simplified_false_val =
+      IRSimplifier::simplify(false_val->accept_mutator(this));
+  if (simplified_condition->isConstant()) {
+    return immediateAs<int>(simplified_condition) ? simplified_true_val
+                                                  : simplified_false_val;
+  }
+
+  bool nothing_changed = (simplified_condition == condition) &&
+      (simplified_true_val == true_val) && (simplified_false_val == false_val);
+  return nothing_changed
+      ? v
+      : alloc<IfThenElse>(
+            simplified_condition, simplified_true_val, simplified_false_val);
+}
+
+ExprPtr SimplifierUnderContext::mutate(CompareSelectPtr v) {
+  GRAPH_DEBUG("(SimplifierUnderContext) Original: ", std::to_string(v));
+
+  ExprPtr lhs = v->lhs();
+  ExprPtr rhs = v->rhs();
+  ExprPtr ret1 = v->ret_val1();
+  ExprPtr ret2 = v->ret_val2();
+
+  auto simplified_lhs = IRSimplifier::simplify(lhs->accept_mutator(this));
+  auto simplified_rhs = IRSimplifier::simplify(rhs->accept_mutator(this));
+  auto simplified_ret1 = IRSimplifier::simplify(ret1->accept_mutator(this));
+  auto simplified_ret2 = IRSimplifier::simplify(ret2->accept_mutator(this));
+
+  ExprPtr simplified_cmp_select_expr = nullptr;
+  if ((simplified_lhs == lhs) && (simplified_rhs == rhs) &&
+      (simplified_ret1 == ret1) && (simplified_ret2 == ret2)) {
+    simplified_cmp_select_expr = v;
+  } else {
+    simplified_cmp_select_expr = alloc<CompareSelect>(
+        simplified_lhs,
+        simplified_rhs,
+        simplified_ret1,
+        simplified_ret2,
+        v->compare_select_op(),
+        v->bias());
+  }
```

Comment:
Why can't we just simplify the entire `CompareSelect` expression instead of doing each of the inner expressions separately?

Reply:
> Why can't we just simplify the entire `CompareSelect` expression instead of doing each of the inner expressions separately?

Please correct me if I don't understand exactly.
- The `IRSimplifier::simplify` also invokes this mutator https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/tensorexpr/ir_simplifier.cpp#L2968. If we simplify the entire `CompareSelect`, it might lead to infinite recursive invocation.
- Besides that, the `CompareSelect` could be nested, so we need to simplify the inner expression to make sure each `CompareSelect` is simplified.

Reply:
Yes, you are right about that. I missed it earlier. This looks good.